Various additions

funkyfuture opened the pull request:
i'm absolutely not sure whether i approached the `XMLParser`'s target argument type correctly. i don't know if it can be addressed at the moment at all, as there is not class provided by lxml that would be used as base for such classes.

funkyfuture committed:
Adds a stub for etree.XMLParser

It comes with a Protocol based typr for a parser target.

fix parserarget

funkyfuture committed:
Adds etree._Element.getparent

funkyfuture committed:
Adds etree._ElementTree.getroot and .xslt

Also reformats .xpath

funkyfuture committed:
Adds etree.QName

funkyfuture commented:
i just noticed that there is `typing_extensions` that could be used for structural typing. @JelleZijlstra would it be feasible to add this as dependency?

JelleZijlstra commented:
Yes, it's fine to use `Protocol` in stubs; the dependency doesn't even have to be declared.

funkyfuture committed:
Adds a type for namespace maps in the etree module

funkyfuture committed:
Fixes some attributes of etree._Element

I included the _attrib as some of its methods return rather
unexptected objects.

funkyfuture committed:
Adds etree._Element.__getitem__, .__len__ and .getnext

funkyfuture committed:
Adds etree._Element.__delitem__

funkyfuture committed:
Adds etree._ElementTree.parser and ._setroot

funkyfuture committed:
Adds etree._Element.getprevious

funkyfuture committed:
Adds etree._Element.makelelement

funkyfuture committed:
Fixes etree._Element.append's signature

funkyfuture committed:
Adds etree._Element.getroottree and .remove

funkyfuture committed:
Adds etree.Comment and etree.PI

funkyfuture committed:
Adds etree._Element.iter

funkyfuture committed:
Adds etree._Eleemnt.insert

zadamah commented:
Hello,
First, thank you @JelleZijlstra for starting and sharing this project.
I was intending to add some type annotations that I needed such as `XMLParser`. That's how I landed here, seeing that @funkyfuture already did that and then more, thank you :raised_hands:.
So, I was just wondering if you will merging these changes soon. Also, I would be very happy to contribute if any help is needed.

JelleZijlstra commented:
I'd be OK with merging but there is a merge conflict. @funkyfuture could you fix it? Thanks for contributing these stubs!

funkyfuture committed:
Adds etree._ElementTree.getpath

funkyfuture commented:
done.

funkyfuture committed:
Add cssselect.CSSSelector and etree._Element.cssselect

funkyfuture committed:
Adds etree._ElementTree.write_c14n

funkyfuture committed:
Adds the base class of etree._XSLTResultTree

funkyfuture committed:
Adds etree.cleanup_namespaces

JelleZijlstra commented on lxml/cssselect.pyi:
```diff
@@ -0,0 +1,15 @@
+    def __init__(self,
+                 css: str,
+                 namespaces: Optional[_DictAnyStr] = None,
+                 translator: str = "xml"): ...
```
Use `...` for all defaults.

JelleZijlstra commented on lxml/cssselect.pyi:
```diff
@@ -0,0 +1,15 @@
+from typing import Dict, List, Optional, Union
```
This stubs seems to be incomplete. Could you add a module getattr (see https://www.python.org/dev/peps/pep-0484/#stub-files) to indicate this?

funkyfuture replied:
i'm also addings this to the other modue stubs.

JelleZijlstra commented on lxml/cssselect.pyi:
```diff
@@ -0,0 +1,15 @@
+
+# _AnyStr = Union[str, bytes]
+# _ListAnyStr = Union[List[str], List[bytes]]
+_DictAnyStr = Union[Dict[str, str], Dict[bytes, bytes]]
```
Can this be a Mapping instead?

funkyfuture replied:
i copied this from the `etree` stub.

i'm not sure whether the `etree._NSMap` type would be appropriate here. for the `xpath` method, which i think is used eventually, it wouldn't as this one doesn't accept a mapping with a default namespace (`None` as key`).

JelleZijlstra commented on lxml/cssselect.pyi:
```diff
@@ -0,0 +1,15 @@
+from lxml import  etree
+
+
+# _AnyStr = Union[str, bytes]
```
Remove these if they're not going to be used.

JelleZijlstra commented on lxml/cssselect.pyi:
```diff
@@ -0,0 +1,15 @@
+from typing import Dict, List, Optional, Union
+
+from lxml import  etree
```
Only one space after "import"

JelleZijlstra commented on lxml/etree.pyi:
```diff
@@ -83,6 +84,11 @@ class _ElementTree:
               inclusive_ns_prefixes: _ListAnyStr = ...) -> None: ...
+    def write_c14n(self,
+                   file: Union[_AnyStr, typing.IO],
+                   with_comments: bool = True,
```
Here too, use `...` for defaults.

funkyfuture committed:
Adds a __getattr__ to all modules to indicate incompleteness

funkyfuture committed:
Cleans imports from typing in etree

funkyfuture committed:
Fixes signatures of etree.Element and .SubElement

funkyfuture committed:
Fixes etree._Element.find's signature

funkyfuture committed:
Fixes etree.QNames.__init__'s signature

funkyfuture committed:
Adds etree._Comment and ._ProcessingInstruction

funkyfuture commented:
@JelleZijlstra is there anything i can do to get this going?

JelleZijlstra committed:
Merge branch 'master' into additions_as_needed
